### PR TITLE
LF-4216 - updated requirements

### PR DESCRIPTION
**Pull request by Duncan-Brain**
**Description**

There have been updated requirements to the soil amendment task.

This reflects those changes

Completed:
- Migrations

TODO
- Everything else

Jira link: [LF-4216](https://lite-farm.atlassian.net/browse/LF-4216)

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [ ] Other (please explain)

**Checklist:**

- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

[LF-4216]: https://lite-farm.atlassian.net/browse/LF-4216?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js [hunk 1]
```diff
@@ -13,101 +13,522 @@
...
+      [],
+      'npk_percent_check',
+    );
+  });
 
   // Reverse product data migration
   await knex.schema.alterTable('soil_amendment_task', (table) => {
     table.decimal('product_quantity', 36, 12);
-    table.enu('product_quantity_unit', productQuantityUnits);
+    table.text('product_quantity_unit');
```
Comment (kathyavini) [inline packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:417]: Why text if it was an enum originally?

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js [hunk 2]
```diff
@@ -13,101 +13,522 @@
...
     table.integer('product_id').references('product_id').inTable('product');
+    table.string('other_purpose');
+    table.text('purpose');
```
Comment (kathyavini) [inline packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:420]: Same. Purpose is an enum on the current tables.

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js (hunk 1 above)
Comment (Duncan-Brain) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:417]: I thought I checked that it is currently text so it should go back to that. Will double check!

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js (hunk 2 above)
Comment (Duncan-Brain) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:420]: Same I was seeing them as text will double check beta see if i messed up my local or something

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js [hunk 3]
```diff
@@ -13,6 +13,20 @@
  *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
  */
 
+const soilAmendmentFertiliserTypeKeys = ['DRY', 'LIQUID'];
+const elementalUnits = ['percent', 'ratio', 'ppm', 'mg/kg'];
+const molecularCompoundsUnits = ['ppm', 'mg/kg'];
+const rawElements = 'n, p, k, calcium, magnesium, sulfur, copper, manganese, boron';
+const rawElementsAddition = 'n + p + k + calcium + magnesium + sulfur + copper + manganese + boron';
```
[the remark this answers is not in the document]
Comment (Duncan-Brain) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:20]: Probably some way to make this with an array and function lol I just haven't taken the time

[the remark this answers is not in the document]
Comment (kathyavini) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:20]: Don't change it; I love the string SQL 😂

File: packages/api/tests/mock.factories.js [hunk 4]
```diff
@@ -1466,25 +1466,44 @@ function fakeDisease(defaultData = {}) {
...
+
+  // The model normally handles this conversion, but this is a direct insert into the database
+  const isVolume = volumeUnits.includes(pestTask.product_quantity_unit);
+  const isWeight = weightUnits.includes(pestTask.product_quantity_unit);
+
+  const updatedPestTask = {
+    ...pestTask,
+    volume: isVolume ? pestTask.product_quantity : null,
+    volume_unit: isVolume ? pestTask.product_quantity_unit : null,
+    weight: isWeight ? pestTask.product_quantity : null,
```
Comment (kathyavini) [inline packages/api/tests/mock.factories.js:1488]: @Duncan-Brain I realized updating this test (yay, testing!) that the stubbed code needs logic to `null` the other property (weight/volume) or else the PATCH can violate the volume or weight check if `product_quantity_unit` switches types. Could you update the stub?

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js (hunk 1 above)
Comment (Duncan-Brain) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:417]: So I am seeing it as currently text on beta defaulting to 'kg' (which I think I forgot here):
Screenshot 2024-05-29 at 9 18 33 AM

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js (hunk 2 above)
Comment (Duncan-Brain) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:420]: Purpose is also currently text as I see it: image

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js (hunk 1 above)
Comment (kathyavini) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:417]: So you can see in the migration that created the table that they are both enumerations, but that's not a datatype -- you'll have to go to `constraints` > check (in the screenshot you have above) to see the list of allowable values.

I don't know how to create this view in pgAdmin4, but in Postico it's easiest to see in the DDL:

Screenshot 2024-05-29 at 7 59 27 AM

Comment (Duncan-Brain) [reply packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:417]: Ah I def thought enum was a type, thanks! Will fix!

File: packages/api/src/models/soilAmendmentProductModel.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    return 'id';
+  }
+
+  // Optional JSON schema. This is not the database schema! Nothing is generated
+  // based on this. This is only used for validation. Whenever a model instance
+  // is created it is checked against this schema. http://json-schema.org/.
+  static get jsonSchema() {
+    return {
+      type: 'object',
+      required: ['key'],
```
Comment (kathyavini) [inline packages/api/src/models/soilAmendmentProductModel.js:32]: Copy-paste error, right? I don't think there is a `key` on this table

File: packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js [hunk 6]
```diff
@@ -13,101 +13,542 @@
  *  GNU General Public License for more details, see <https://www.gnu.org/licenses/>.
  */
 
-// No changes here to maintain compatibility with the original data
-const productQuantityUnits = ['g', 'lb', 'kg', 't', 'mt', 'oz', 'l', 'gal', 'ml', 'fl-oz'];
+const soilAmendmentFertiliserTypeKeys = ['DRY', 'LIQUID'];
+const elementalUnits = ['percent', 'ratio', 'ppm', 'mg/kg'];
+const molecularCompoundsUnits = ['ppm', 'mg/kg'];
+const elements = ['n', 'p', 'k', 'calcium', 'magnesium', 'sulfur', 'manganese', 'boron'];
```
Comment (Duncan-Brain) [inline packages/api/db/migration/20240504185218_create_soil_amendement_task_products_table.js:19]: @kathyavini dont hate me! I simplified it haha

File: packages/api/tests/mock.factories.js (hunk 4 above)
Comment (Duncan-Brain) [reply packages/api/tests/mock.factories.js:1522]: Should be done!

**Pull request merged**